core/vmenv.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/ethereumproject/go-ethereum/core/vm"
+	"github.com/ethereumproject/go-ethereum/machine/classic"
+)
+
+type VmEnv interface {
+	Coinbase() common.Address
+	BlockNumber() *big.Int
+	RuleSet() vm.RuleSet
+	Db() *state.StateDB
+	Call(sender common.Address, to common.Address, data []byte, gas, price, value *big.Int) ([]byte, error)
+	Create(me state.AccountObject, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error)
+}
+
+type vmEnv struct {
+	coinbase 	common.Address
+	number  	*big.Int
+	db      	*state.StateDB
+	hashfn		func(n uint64) common.Hash
+	fork		vm.Fork
+	rules       vm.RuleSet
+	machine		vm.Machine
+}
+
+func (self *vmEnv) Coinbase() common.Address {
+	return self.coinbase
+}
+
+func (self *vmEnv) BlockNumber() *big.Int {
+	return self.number
+}
+
+func (self *vmEnv) RuleSet() vm.RuleSet {
+	return self.rules
+}
+
+func (self *vmEnv) Db() *state.StateDB {
+	return self.db
+}
+
+func (self *vmEnv) do(callOrCreate func(*vmEnv)(vm.Context,error)) ([]byte, common.Address, error) {
+	for {
+		var (
+			context vm.Context
+			err error
+			out []byte
+			address common.Address
+		)
+		context, err = callOrCreate(self)
+		if err == nil {
+			out, address, err = self.execute(context)
+			context.Finish()
+		}
+		if err != machineBrokenError {
+			return out, address, err
+		} else {
+			// ?? mybe will better to switch to the classic vm and try again before ??
+			panic(err)
+		}
+	}
+}
+
+func (self *vmEnv) Call(sender common.Address, to common.Address, data []byte, gas, price, value *big.Int) ([]byte, error) {
+	out, _, err := self.do(func(e *vmEnv)(vm.Context,error){
+		return e.machine.Call(e.number.Uint64(),sender,to,data,gas,price,value)
+	})
+	return out, err
+}
+
+func (self *vmEnv) Create(caller state.AccountObject, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error) {
+	return self.do(func(e *vmEnv)(vm.Context,error){
+		return e.machine.Create(e.number.Uint64(),caller.Address(),data,gas,price,value)
+	})
+}
+
+var machineBrokenError = errors.New("machine broken")
+
+func (self *vmEnv) execute(ctx vm.Context) ([]byte,common.Address,error) {
+	for {
+		req := ctx.Fire()
+		if req != nil {
+			switch req.ID {
+			case vm.RequireAccount:
+				acc := self.db.GetAccount(req.Address)
+				addr := acc.Address()
+				nonce := acc.Nonce()
+				balance := acc.Balance()
+				ctx.CommitAccount(addr,nonce,balance)
+			case vm.RequireCode:
+				addr := req.Address
+				code := self.db.GetCode(addr)
+				hash := self.db.GetCodeHash(addr)
+				ctx.CommitCode(addr,hash,code)
+			case vm.RequireHash:
+				number := req.Number
+				hash := self.hashfn(number)
+				ctx.CommitBlockhash(number,hash)
+			case vm.RequireRules:
+				number := new(big.Int).SetUint64(req.Number)
+				fork := self.fork
+				ctx.CommitRules(req.Number,self.rules.GasTable(number),fork)
+			default:
+				if ctx.Status() == vm.RequireErr {
+					// ?? unsupported VM implementaion ??
```

Comment:
If it's unsupported implementation, then the client should just panic.

This is practically impossible to happen and will only be possible for some hand-crafted invalid transactions/blocks.

But are you should RequireError is "unsupported VM implementation"? I think it should be a separate error.